application/client/src/app/ui/tabs/sources/multiplefiles/structure/component.ts

```diff
@@ -0,0 +1,191 @@
+import { FileHolder } from '../file.holder';
+import { MatTable } from '@angular/material/table';
+import { bytesToStr } from '@env/str';
+
+export const COLUMNS = {
+    type: 'type',
+    name: 'name',
+    path: 'path',
+    size: 'size',
+    modification_date: 'modification date',
+};
+
+@Component({
+    selector: 'app-tabs-source-multiplefiles-structure',
+    templateUrl: './template.html',
+    styleUrls: ['./styles.less'],
+    encapsulation: ViewEncapsulation.None,
+})
+@Ilc()
+export class TabSourceMultipleFilesStructure
+    extends ChangesDetector
+    implements AfterContentInit, AfterViewInit
+{
+    columns: string[] = [
+        COLUMNS.type,
+        COLUMNS.name,
+        COLUMNS.path,
+        COLUMNS.size,
+        COLUMNS.modification_date,
+    ];
+
+    @Input() files!: FileHolder[];
+    @Output() context: EventEmitter<null | FileHolder[]> = new EventEmitter();
+
+    @ViewChild(MatSort) sort!: MatSort;
+    @ViewChild(MatTable) table!: MatTable<any>;
+
+    public data!: MatTableDataSource<FileHolder>;
+
+    constructor(cdRef: ChangeDetectorRef) {
+        super(cdRef);
+    }
+
+    public ngAfterContentInit() {
+        this.data = new MatTableDataSource<FileHolder>(this.files);
+    }
+
+    public ngAfterViewInit() {
+        this.data.sort = this.sort;
+    }
+
+    public ngModificationDate(time: number): string {
+        return new Date(time).toLocaleString('en-GB');
+    }
+
+    public ngOnSortChange() {
+        this.detectChanges();
+        this.table.renderRows();
+    }
+
+    public ngOnRowSelect(file: FileHolder) {
+        file.reverseSelect();
+        this.context.next(null);
+    }
+
+    public ngSize(size: number): string {
+        return bytesToStr(size);
+    }
+
+    public ngContextMenu(event: any, file?: FileHolder) {
+        event.stopImmediatePropagation();
+        const items = [];
+        if (file !== undefined) {
+            items.push({
+                caption: file.selected ? 'Unselect' : 'Select',
+                handler: () => {
+                    file.selected ? file.unselect() : file.select();
+                    this.context.next(null);
+                },
+            });
+        }
+        items.push(
+            {
+                caption: 'Select All',
+                handler: () => {
+                    this.data.data.forEach((f: FileHolder) => f.select());
+                    this.context.next(null);
+                },
+            },
+            {
+                caption: 'Unselect All',
+                handler: () => {
+                    this.data.data.forEach((f: FileHolder) => f.unselect());
+                    this.context.next(null);
+                },
+            },
+            {
+                caption: 'Reverse Select All',
+                handler: () => {
+                    this.data.data.forEach((f: FileHolder) => f.reverseSelect());
+                    this.context.next(null);
+                },
+            },
+        );
```

Comment:
you are welcome to use delimiters in the menu to make it more organized, just add an empty object
```
items.push(
            {
                caption: 'Select All',
                handler: () => { ... },
            },
            {
                caption: 'Unselect All',
                handler: () => { ... },
            },
            { /* This is delimiter */},
            {
                caption: 'Reverse Select All',
                handler: () => { ... },
            },
        );
```
I don't ask to put a delimiter like in the example, just to show how you can do it. You can add delimiters between groups of commands. Like selections items, removing items, and so on.